File: apps/client/src/mocks/fixtures/practice.ts
import type {
  PracticeQuestionTurnResponse,
  PracticeResponse,
  PracticeResultResponse,
  TaskFailureResponse,
} from "@/api/generated/models"
import { toPracticeQuestion, toPracticeSession } from "@/models/practice-response"
import type { PracticeSelection } from "@/models/practice-workflow"

export const practiceQuestionFixture: PracticeQuestionTurnResponse = {
  id: "10000000-0000-4000-8000-000000000003",
  role: "assistant",

  content: "请结合一个真实经历，说明你如何分析问题、做出关键判断、推动行动并验证最终结果。",
  criteria: [
    { dimension: "个人贡献", expectation: "说明你亲自采取的关键行动及其作用。" },
    { dimension: "结果与证据", expectation: "提供可验证的结果，并说明验证依据。" },
    { dimension: "技术决策", expectation: "说明方案取舍、关键判断及其依据。" },
  ],
  guidance: {
    hints: ["说明你的关键判断和验证依据。"],
    framework: ["背景、目标、行动与结果"],
  },
  referenceAnswer: "我先明确目标和约束，再用小范围验证选择方案，推动落地后对比结果并复盘。",
}

export const practiceFollowUpFixture: PracticeQuestionTurnResponse = {
  id: "10000000-0000-4000-8000-000000000005",
  role: "assistant",

  content: "你如何验证结果来自你的关键行动，还有哪些因素可能影响结论？",
  guidance: {
    hints: ["说明对照证据和结论边界。"],
    framework: ["验证方法、证据与局限"],
  },
  referenceAnswer: "我会对比行动前后的结果，排查同期变化，并说明证据能支持的结论范围。",

  criteria: [],
}

export const practiceResultFixture: PracticeResultResponse = {
  score: 78,
  dimensionScores: {
    relevance: { score: 82, explanation: "回答基本围绕问题展开。" },
    structure: { score: 80, explanation: "背景、行动和结果的顺序清楚。" },
    specificity: { score: 82, explanation: "能够说明具体的实施行动。" },
    contribution: { score: 79, explanation: "说明了个人负责的关键行动。" },
    evidence: { score: 62, explanation: "提及结果，但缺少明确的对照证据。" },
    roleAlignment: { score: 83, explanation: "所述经历与岗位要求相关。" },
    communication: { score: 80, explanation: "文字表达清楚，信息衔接自然。" },
    riskAwareness: { score: 76, explanation: "考虑了约束，但风险边界还可展开。" },
  },
  summary: "回答清楚地呈现了行动过程，可以进一步补充结果证据。",
  strengths: ["能够说明个人采取的关键行动。"],
  issues: ["结果的验证依据不够具体。"],
  suggestions: ["补充一个可比较的结果指标。"],
}

export const practiceResponseFixture: PracticeResponse = {
  id: "10000000-0000-4000-8000-000000000001",
  role: {
    id: "10000000-0000-4000-8000-000000000010",
    title: "Senior Frontend Engineer",
    company: "ByteDance",
  },
  questionType: "project",
  difficulty: "basic",
  rounds: [
    {
      id: "10000000-0000-4000-8000-000000000002",
      sequence: 0,
      turns: [
        practiceQuestionFixture,
        {
          id: "10000000-0000-4000-8000-000000000004",
          role: "user",
          content: "我明确了目标和约束，比较方案后推动落地，并用结果验证判断。",
        },
        practiceFollowUpFixture,
        {
          id: "10000000-0000-4000-8000-000000000006",
          role: "user",
          content: "我对比了行动前后的指标，并排查同期变化带来的影响。",
        },
      ],
      result: practiceResultFixture,
    },
  ],
  createdAt: "2026-09-18T08:00:00Z",
  endedAt: null,
}

const review = toPracticeSession(practiceResponseFixture, { status: "idle", error: null })
if (review.status !== "review") throw new Error("Practice response fixture must contain a result.")

export const practiceTaskFailureInput = "__RIVA_MOCK_PRACTICE_TASK_FAILURE__"
export const practiceTaskFailureFixture: TaskFailureResponse = {
  status: "failed",
  error: { code: "llm_unavailable", message: "LLM service is temporarily unavailable." },
}

// Display samples for isolated component tests and stories. Runtime mocks use the API fixtures.
export const practiceFixture = {
  context: review.context,
  completion: { questionsCompleted: 1, finalAttemptAverageScore: practiceResultFixture.score },
  selection: {
    roleId: null,
    questionType: "project",
    difficulty: "basic",
  } satisfies PracticeSelection,
  question: toPracticeQuestion(practiceQuestionFixture),
  followUp: { question: toPracticeQuestion(practiceFollowUpFixture) },
  evaluation: review.evaluation,
  review: review.review,
}

File: apps/server/src/riva/models/practice.py
from datetime import datetime
from enum import StrEnum
from typing import TYPE_CHECKING, Annotated, Literal
from uuid import UUID, uuid4

from pydantic import BaseModel, Field
from sqlalchemy import (
    DateTime,
    Enum,
    ForeignKey,
    String,
    UniqueConstraint,
    text,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql.sqltypes import Uuid

from riva.db.base import Base
from riva.db.types import PydanticJSONB
from riva.models.career_profile import CareerProfileContent
from riva.models.mixins import TaskStateMixin
from riva.models.role import RoleContent
from riva.models.types import NonBlankStr
from riva.utils import utc_now

if TYPE_CHECKING:
    from riva.models.role import Role


class PracticeQuestionType(StrEnum):
    PROJECT = "project"
    BEHAVIORAL = "behavioral"
    BUSINESS_UNDERSTANDING = "business_understanding"
    MOTIVATION = "motivation"
    TECHNICAL_BASICS = "technical_basics"


class PracticeDifficulty(StrEnum):
    BASIC = "basic"
    HARD = "hard"


class PracticeGuidance(BaseModel):
    hints: list[NonBlankStr] = Field(
        description="Concrete points to help the candidate address this question."
    )
    framework: list[NonBlankStr] = Field(
        description="Ordered steps for organizing an answer to this question."
    )


class PracticeCriterion(BaseModel):
    dimension: NonBlankStr = Field(
        description="Question-specific assessment area, not a fixed scoring dimension."
    )
    expectation: NonBlankStr = Field(
        description="What a satisfactory answer should demonstrate in this assessment area."
    )


class PracticeQuestionTurn(BaseModel):
    id: UUID
    role: Literal["assistant"] = "assistant"
    content: NonBlankStr
    guidance: PracticeGuidance
    criteria: list[PracticeCriterion]
    reference_answer: NonBlankStr


class PracticeAnswerTurn(BaseModel):
    id: UUID
    role: Literal["user"] = "user"
    content: NonBlankStr


type PracticeTurnContent = Annotated[
    PracticeQuestionTurn | PracticeAnswerTurn,
    Field(discriminator="role"),
]


class PracticeDimensionScore(BaseModel):
    score: int = Field(
        strict=True,
        ge=0,
        le=100,
        description=(
            "90-100: complete, clear and concretely supported; "
            "70-89: basically meets requirements with minor gaps; "
            "50-69: partially meets requirements but lacks key information; "
            "1-49: severely insufficient with limited relevant evidence; "
            "0: entirely fails requirements or provides no relevant evidence."
        ),
    )
    explanation: NonBlankStr = Field(
        description="Specific reasons for the score, grounded in the candidate's actual answers and remaining gaps."
    )


class PracticeDimensionScores(BaseModel):
    relevance: PracticeDimensionScore = Field(
        description="addresses the question's core requirements."
    )
    structure: PracticeDimensionScore = Field(
        description="connects context, actions, decisions and outcomes coherently."
    )
    specificity: PracticeDimensionScore = Field(
        description="provides concrete scenarios, technical details and precise descriptions."
    )
    contribution: PracticeDimensionScore = Field(
        description="distinguishes personal responsibility and actions from team work."
    )
    evidence: PracticeDimensionScore = Field(
        description="supports outcomes with facts, measurements, business value or verification."
    )
    role_alignment: PracticeDimensionScore = Field(
        description="demonstrates abilities relevant to the target role and JD."
    )
    communication: PracticeDimensionScore = Field(
        description="uses clear, natural language with useful information density (written answers only, not unobserved vocal delivery)."
    )
    risk_awareness: PracticeDimensionScore = Field(
        description="considers relevant risks, failure cases and boundaries."
    )


class PracticeResult(BaseModel):
    score: float = Field(ge=0, le=100, allow_inf_nan=False)
    dimension_scores: PracticeDimensionScores
    summary: NonBlankStr = Field(
        description="Overall assessment grounded in dimension scores, their explanations, actual answers and question criteria."
    )
    strengths: list[NonBlankStr] = Field(
        description="Strengths from dimensions scoring above 70 and their explanations, supported by actual answers; empty if none."
    )
    issues: list[NonBlankStr] = Field(
        description="Issues from dimensions scoring below 70 and their explanations, supported by actual answers; empty if none."
    )
    suggestions: list[NonBlankStr] = Field(
        description="Specific actionable improvements for the identified issues, grounded in question criteria and actual answers."
    )


class PracticeSession(Base):
    __tablename__ = "practice_sessions"

    id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True), primary_key=True, default=uuid4
    )
    user_id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), index=True
    )
    role_id: Mapped[UUID | None] = mapped_column(
        Uuid(as_uuid=True), ForeignKey("roles.id", ondelete="SET NULL"), index=True
    )
    # Captured at practice creation and refreshed immediately before role deletion.
    role_title_snapshot: Mapped[str] = mapped_column(String)
    role_company_snapshot: Mapped[str | None] = mapped_column(String)
    role: Mapped[Role | None] = relationship(lazy="selectin")
    # Immutable AI context, independent of the live role and display snapshots.
    profile_snapshot: Mapped[CareerProfileContent] = mapped_column(
        PydanticJSONB(CareerProfileContent)
    )
    role_snapshot: Mapped[RoleContent] = mapped_column(PydanticJSONB(RoleContent))
    question_type: Mapped[PracticeQuestionType] = mapped_column(
        Enum(
            PracticeQuestionType,
            values_callable=lambda enum: [member.value for member in enum],
            native_enum=False,
            create_constraint=True,
            validate_strings=True,
            name="practice_question_type",
        )
    )
    difficulty: Mapped[PracticeDifficulty] = mapped_column(
        Enum(
            PracticeDifficulty,
            values_callable=lambda enum: [member.value for member in enum],
            native_enum=False,
            create_constraint=True,
            validate_strings=True,
            name="practice_difficulty",
        )
    )
    max_follow_ups: Mapped[int]
    ended_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utc_now
    )
    rounds: Mapped[list[PracticeRound]] = relationship(
        cascade="all, delete-orphan",
        lazy="selectin",
        order_by="PracticeRound.sequence",
    )


class PracticeRound(TaskStateMixin, Base):
    __tablename__ = "practice_rounds"
    __table_args__ = (
        # Replacement rounds briefly share a sequence while the main turn is moved.
        UniqueConstraint(
            "practice_id",
            "sequence",
            name="practice_round_sequence_unique",
            deferrable=True,
            initially="DEFERRED",
        ),
    )

    id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True), primary_key=True, default=uuid4
    )
    practice_id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True), ForeignKey("practice_sessions.id", ondelete="CASCADE")
    )
    sequence: Mapped[int]
    result: Mapped[PracticeResult | None] = mapped_column(PydanticJSONB(PracticeResult))
    turns: Mapped[list[PracticeTurn]] = relationship(
        cascade="all, delete-orphan",
        lazy="selectin",
        order_by="PracticeTurn.sequence",
    )


class PracticeTurn(Base):
    __tablename__ = "practice_turns"
    __table_args__ = (
        UniqueConstraint("round_id", "sequence", name="practice_turn_sequence_unique"),
    )

    id: Mapped[UUID] = mapped_column(Uuid(as_uuid=True), primary_key=True)
    round_id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True), ForeignKey("practice_rounds.id", ondelete="CASCADE")
    )
    sequence: Mapped[int]
    role: Mapped[str] = mapped_column(
        Enum(
            "assistant",
            "user",
            native_enum=False,
            create_constraint=True,
            validate_strings=True,
            name="practice_turn_role",
        )
    )
    content: Mapped[str] = mapped_column(String)
    guidance: Mapped[PracticeGuidance | None] = mapped_column(
        PydanticJSONB(PracticeGuidance)
    )
    criteria: Mapped[list[PracticeCriterion]] = mapped_column(
        PydanticJSONB(list[PracticeCriterion]),
        default=list,
        server_default=text("'[]'::jsonb"),
    )
    reference_answer: Mapped[str | None] = mapped_column(String)
